pyface/data_view/i_data_view_widget.py

```diff
@@ -8,52 +8,354 @@
+
+    # ------------------------------------------------------------------------
+    # MDataViewWidget Interface
+    # ------------------------------------------------------------------------
+
+    def _header_visible_updated(self, event):
+        """ Observer for header_visible trait. """
+        if self.control is not None:
+            self._set_control_header_visible(event.new)
+
+    def _get_control_header_visible(self):
+        """ Toolkit specific method to get the visibility of the header.
+
+        Returns
+        -------
+        control_header_visible : bool
+            Whether or not the control's header is visible.
+        """
+        raise NotImplementedError()
+
+    def _set_control_header_visible(self, control_header_visible):
+        """ Toolkit specific method to set the visibility of the header.
+
+        Parameters
+        ----------
+        control_header_visible : bool
+            Whether or not the control's header is visible.
+        """
+        raise NotImplementedError()
+
+    def _selection_type_updated(self, event):
+        """ Observer for selection_type trait. """
+        if self.control is not None:
+            self._set_control_selection_type(event.new)
+            self.selection = []
+
+    def _get_control_selection_type(self):
+        """ Toolkit specific method to get the selection type.
+
+        Returns
+        -------
+        selection_type : str
+            The type of selection the control is using.
+        """
+        raise NotImplementedError()
+
+    def _set_control_selection_type(self, selection_type):
+        """ Toolkit specific method to change the selection type.
+
+        Parameters
+        ----------
+        selection_type : str
+            The type of selection the control is using.
+        """
+        raise NotImplementedError()
+
+    def _selection_mode_updated(self, event):
+        """ Observer for selection_mode trait. """
+        if self.control is not None:
+            self._set_control_selection_mode(event.new)
+            self.selection = []
+
+    def _get_control_selection_mode(self):
+        """ Toolkit specific method to get the selection mode.
+
+        Returns
+        -------
+        selection_mode : str
+            The selection mode the control is using (eg. single vs. extended
+            selection).
+        """
+        raise NotImplementedError()
+
+    def _set_control_selection_mode(self, selection_mode):
+        """ Toolkit specific method to change the selection mode.
+
+        Parameters
+        ----------
+        selection_mode : str
+            The selection mode the control is using (eg. single vs. extended
+            selection).
+        """
+        raise NotImplementedError()
+
+    def _selection_updated(self, event):
+        """ Observer for selection trait. """
+        if self.control is not None and not self._selection_updating_flag:
```

Comment:
Just to clarify... so is this a "just-in-case" kind of things or do we actually have a loop-back issue?

This kind of internal facing defensive programming costs readability, maintainability and may mask bugs: It is hard to remove / refactor these guards when it gets in the way in the future or it actually causes some other bugs. In practice, no one dares to remove these guards even if they are actually not needed, and if these guards are causing bugs, one ends up workaround because it is hard to know whether the guard is still needed. I understand and agree with the concern with robustness, just wanted to mention the tradeoff here. 

If this is a just-in-case kind of thing, a comment to explain this would at least mitigate some of that readability cost.

Reply:
> If this is a just-in-case kind of thing, a comment to explain this would at least mitigate some of that readability cost.

Actually, if we are keeping the `"items"` in the observers, then this is not a just-in-case thing. The implementation does go into an infinite loop with setting an empty list.